[KOGITO-4602] - Create Go module for product kogito operator

vaibhavjainwiz opened the pull request:
Many thanks for submitting your Pull Request :heart:! 

Jira issue : https://issues.redhat.com/browse/KOGITO-4602

Please make sure your PR meets the following requirements:

- [ ] You have read the [contributors' guide](https://github.com/kiegroup/kogito-cloud-operator/blob/master/README.md#contributing-to-the-kogito-operator)
- [ ] Pull Request title is properly formatted: `[KOGITO-XYZ] Subject`
- [ ] Pull Request contains a link to the JIRA issue
- [ ] Pull Request contains a description of the issue
- [ ] Pull Request does not include fixes for issues other than the main ticket
- [ ] Your feature/bug fix has a unit test that verifies it
- [ ] You've tested the new feature/bug fix in an actual OpenShift cluster
- [ ] You've added a [RELEASE_NOTES.md](RELEASE_NOTES.md) entry regarding this change

ricardozanini commented on Makefile:
```diff
@@ -1,9 +1,9 @@
 # Current Operator version
-VERSION ?= 2.0.0-snapshot
+VERSION ?= 1.0.0-snapshot
```
Product version: 7.11

cc @radtriste @spolti

vaibhavjainwiz replied:
changed to 7.11

ricardozanini commented on bundle/manifests/rhpam.kiegroup.org_kogitobuilds.yaml:
```diff
@@ -4,9 +4,9 @@ metadata:
     controller-gen.kubebuilder.io/version: v0.3.0
   creationTimestamp: null
-  name: kogitobuilds.app.kiegroup.org
+  name: kogitobuilds.rhpam.kiegroup.org
```
Can be deleted.

ricardozanini commented on config/crd/bases/rhpam.kiegroup.org_kogitobuilds.yaml:
```diff
@@ -6,9 +6,9 @@ metadata:
     controller-gen.kubebuilder.io/version: v0.3.0
   creationTimestamp: null
-  name: kogitobuilds.app.kiegroup.org
+  name: kogitobuilds.rhpam.kiegroup.org
```
Can be deleted.

ricardozanini commented on config/manager/manager.yaml:
```diff
@@ -29,9 +29,11 @@ spec:
+            memory: 30Mi
+          requests:
+            cpu: 100m
+            memory: 20Mi
```
I don't think we want this added. Might break the operator pod in certain environments. Let the users decide when to set this configuration.

ricardozanini commented on config/manifests/bases/rhpam-kogito-operator.clusterserviceversion.yaml:
```diff
@@ -0,0 +1,65 @@
+apiVersion: operators.coreos.com/v1alpha1
```
We are missing a lot of information in this file.

vaibhavjainwiz replied:
updated it now. please recheck

ricardozanini commented on go.mod:
```diff
@@ -1,47 +1,29 @@
 	github.com/keycloak/keycloak-operator v0.0.0-20200917060808-9858b19ca8bf
-	github.com/machinebox/graphql v0.2.2
-	github.com/matryer/is v1.4.0 // indirect
+	github.com/kiegroup/kogito-cloud-operator v0.0.3
```
```suggestion
	github.com/kiegroup/kogito-operator v1.4.0
```
After your renaming, right?

vaibhavjainwiz replied:
The tag that we create after renaming of `kogito-operator`, I will replace it here

ricardozanini commented on hack/install.sh:
```diff
@@ -17,12 +17,12 @@
 echo "....... Installing Kogito Operator ${VERSION} ......."
 
-declare url="https://github.com/kiegroup/kogito-cloud-operator/releases/download/${VERSION}/kogito-operator.yaml"
+declare url="https://github.com/kiegroup/rhpam-kogito-operator/releases/download/${VERSION}/kogito-operator.yaml"
```
I think we can get rid of this file since the official installation method would be via OLM.

vaibhavjainwiz replied:
removed this file

ricardozanini commented on version/version.go:
```diff
@@ -16,5 +16,5 @@ package version
 var (
 	// Version of Kogito Operator
-	Version = "2.0.0-snapshot"
+	Version = "1.0.0-snapshot"
```
7.11?

vaibhavjainwiz replied:
changed to 7.11

ricardozanini commented on api/v1/kogitoservices.go:
```diff
@@ -334,11 +330,11 @@ func (k *KogitoServiceSpec) IsInsecureImageRegistry() bool { return k.InsecureIm
 // AddInfra ...
 func (k *KogitoServiceSpec) AddInfra(name string) {
-	k.Infra = append(k.Infra, name)
+	panic("Operation not supported")
```
Raising panic here will crash the operator pod.

vaibhavjainwiz replied:
yes, I added them intentionally. These KogitoInfra related functions like `AddInfra` should not be called in any scenerio. If it is called then we need to fix that calling in code.

sutaakar replied:
Wouldn't it be better to log the error instead of panic?
This way any deployment in any namespace can bring the cluster wide operator down in production.

spolti replied:
I agree, even if user reach this point unintentionally it might crash operator leading to start the "restart loop" until user don't remove his CR, right?
maybe throw a error might be more sense

ricardozanini reviewed (commented):
I did a quick review, please disregard my KogitoBuild comments since we agreed to support it for dev purposes.

Some files that I think we must preserve:

- the LICENSE file
- the README file with some information about this repository and a link to `kogito-operator`
- RELEASE_NOTES should be preserved with a link to upstream. Only `RHPAM` jiras should be added
- examples/example-custom-build-images.yaml 
- examples/quarkus-jvm.Dockerfile and examples/springboot.Dockerfile. These files help users to get started on building their own images. Common use case for enterprise users.
- I didn't understand why we need to remove examples/service-labels.yaml 
- SpringBoot runtimes is out of question here? examples/springboot.yaml 
- I think we need to keep go.tools.mod since we will still use some tooling in this repo.

Thanks for the hard work man!

sutaakar commented on .github/bot-files/comments.yml:
```diff
@@ -2,7 +2,7 @@
 # please see the example for the file below
 
-prFirstTimeContributor: Welcome to the Kogito Community!. Please make sure you've read the [contributors' guide](https://github.com/kiegroup/kogito-cloud-operator/blob/master/README.md#contributing-to-the-kogito-operator)
+prFirstTimeContributor: Welcome to the RHPAM Kogito Community!. Please make sure you've read the [contributors' guide](https://github.com/kiegroup/kogito-cloud-operator/blob/master/README.md#contributing-to-the-kogito-operator)
```
```suggestion
prFirstTimeContributor: Welcome to the RHPAM Kogito Community!. Please make sure you've read the [contributors' guide](https://github.com/kiegroup/kogito-operator/blob/master/README.md#contributing-to-the-kogito-operator)
```

vaibhavjainwiz replied:
I waiting for rename PR to get merge to rename `kogito-cloud-operator` word with `kogito-operator`. I am adding RHPAM in community title

vaibhavjainwiz replied:
done

sutaakar commented on Dockerfile:
```diff
@@ -0,0 +1,27 @@
+# Build the manager binary
```
We should use CEKit instead of direct Dockerfile, as done in https://github.com/kiegroup/kogito-operator/pull/784

vaibhavjainwiz replied:
done

vaibhavjainwiz commented:
> I did a quick review, please disregard my KogitoBuild comments since we agreed to support it for dev purposes.
> 
> Some files that I think we must preserve:
> 
> * the LICENSE file
> * the README file with some information about this repository and a link to `kogito-operator`
> * RELEASE_NOTES should be preserved with a link to upstream. Only `RHPAM` jiras should be added
> * examples/example-custom-build-images.yaml
> * examples/quarkus-jvm.Dockerfile and examples/springboot.Dockerfile. These files help users to get started on building their own images. Common use case for enterprise users.
> * I didn't understand why we need to remove examples/service-labels.yaml
> * SpringBoot runtimes is out of question here? examples/springboot.yaml
> * I think we need to keep go.tools.mod since we will still use some tooling in this repo.
> 
> Thanks for the hard work man!

I had removed these all files because I take approach where first I remove all files from this git repo and then added only necessary file. 
I had added back all files which you had mentioned above

sutaakar commented on api/v1/groupversion_info.go:
```diff
@@ -12,10 +12,26 @@
+/*
+
+
+Licensed under the Apache License, Version 2.0 (the "License");
```
Why is this second licence added here?

vaibhavjainwiz replied:
not sure, I had removed second licence manually

sutaakar commented on bundle.Dockerfile:
```diff
@@ -3,7 +3,7 @@ FROM scratch
 LABEL operators.operatorframework.io.bundle.metadata.v1=metadata/
-LABEL operators.operatorframework.io.bundle.package.v1=kogito-operator
+LABEL operators.operatorframework.io.bundle.package.v1=rhpam-kogito-operator
 LABEL operators.operatorframework.io.bundle.channels.v1=alpha,1.x
```
What OLM channels we should have for the product?

ricardozanini replied:
That's a good question and well spotted. We should have the 7.x only for now, IMO. As soon as we need more, we can add them in the future. I'd avoid too many.

vaibhavjainwiz replied:
set channel to 7.x

sutaakar commented on bundle/manifests/rhpam-kogito-operator.clusterserviceversion.yaml:
```diff
@@ -0,0 +1,336 @@
+  maturity: beta
+  provider:
+    name: Red Hat
+  replaces: kogito-operator.v1.0.0
```
Can be removed

vaibhavjainwiz replied:
removed

sutaakar commented on config/manifests/bases/rhpam-kogito-operator.clusterserviceversion.yaml:
```diff
@@ -0,0 +1,64 @@
+  - name: Documentation
+    url: https://github.com/kiegroup/kogito-cloud-operator/blob/master/README.md
+  - name: Source
+    url: https://github.com/kiegroup/kogito-cloud-operator/
```
```suggestion
    url: https://github.com/kiegroup/rhpam-kogito-operator/
```

vaibhavjainwiz replied:
done

sutaakar commented on controllers/kogitobuild_controller_test.go:
```diff
@@ -1,167 +0,0 @@
-// Copyright 2020 Red Hat, Inc. and/or its affiliates
```
I think we can keep this test as kogitobuild_controller.go exists.

ricardozanini commented:
> I had removed these all files because I take approach where first I remove all files from this git repo and then added only necessary file.

Try to avoid this, please. We want to preserve the authoring of all our files that came from upstream.

sutaakar commented on controllers/kogitoruntime_controller_test.go:
```diff
@@ -1,147 +0,0 @@
-// Copyright 2020 Red Hat, Inc. and/or its affiliates
```
Same here, we should keep the test.

sutaakar commented on example/quarkus-jvm.Dockerfile:
```diff
@@ -8,7 +8,13 @@ ENV RUNTIME_TYPE quarkus
 #   4.1 Optionally test the image locally: docker run --rm -it -p 8080:8080 quay.io/<yournamespace>/process-quarkus-example:latest
 # 5. Push it: docker push quay.io/<yournamespace>/process-quarkus-example:latest
-# 6. Deploy it on Kubernetes with the kogito operator, as a reference use process-quarkus-example-runtime.yaml
+# 6. Deploy it on Openshift with the kogito operator, as a reference use process-quarkus-example-runtime.yaml
```
Do we support just OpenShift for RHPAM operator or also other Kubernetes implementations?

ricardozanini replied:
OKE as well, looks like.

vaibhavjainwiz replied:
Does OKE means `Oracle Cloud Infrastructure Container Engine`?

ricardozanini replied:
https://docs.openshift.com/container-platform/4.3/welcome/oke_about.html

spolti replied:
@sutaakar I think the idea is to support other k8s implementations.

ricardozanini reviewed (approved):
Seems fine, I think we can start running the tests. @sutaakar @radtriste wdyt?

sutaakar commented:
Right, they will just need adjustments...

vaibhavjainwiz commented:
I had tested some sceneios of KogitoBuild and KogitoRuntime locally on crc. They are working fine.

spolti commented on version/version.go:
```diff
@@ -16,5 +16,5 @@ package version
 var (
 	// Version of Kogito Operator
-	Version = "2.0.0-snapshot"
+	Version = "7.11.0"
```
we still discussing it, see the Kogito Cloud channel on gchat.

vaibhavjainwiz replied:
Currently I had initialize it with 7.11.0 as suggested by Zanini. In below comment. We could change it once we fialized it in gchat conversation

spolti commented on Makefile:
```diff
@@ -1,18 +1,18 @@
+CATALOG_IMG ?= quay.io/kiegroup/rhpam-kogito-operator-catalog:$(VERSION)
 # Options for 'bundle-build'
-CHANNELS=alpha,1.x
+CHANNELS=7.x
```
I think we will continue to use 1.x;  let's see.

vaibhavjainwiz replied:
same here.
https://github.com/kiegroup/rhpam-kogito-operator/pull/2#discussion_r595336240

spolti commented on example/springboot.Dockerfile:
```diff
@@ -8,7 +8,7 @@ ENV RUNTIME_TYPE springboot
 #   4.1 Optionally test the image locally: docker run --rm -it -p 8080:8080 quay.io/<yournamespace>/process-springboot-example:latest
 # 5. Push it: docker push quay.io/<yournamespace>/process-springboot-example:latest
-# 6. Deploy it on Kubernetes with the Kogito Operator, as a reference use process-quarkus-example-runtime.yaml (works for both runtimes)
+# 6. Deploy it on Openshift with the Kogito Operator, as a reference use process-quarkus-example-runtime.yaml (works for both runtimes)
```
OpenShift.

vaibhavjainwiz replied:
done

spolti commented on meta/meta.go:
```diff
@@ -18,8 +18,8 @@ import (
-	"github.com/kiegroup/kogito-cloud-operator/api/v1beta1"
-	kafkabetav1 "github.com/kiegroup/kogito-cloud-operator/core/infrastructure/kafka/v1beta1"
+	kafkabetav1 "github.com/kiegroup/kogito-operator/core/infrastructure/kafka/v1beta1"
+	v1 "github.com/kiegroup/rhpam-kogito-operator/api/v1"
```
wouldn't better rename this import to rhpamv1 or something similar for better readability?

vaibhavjainwiz replied:
done. renamed them to rhpamv1

vaibhavjainwiz commented:
> > I had removed these all files because I take approach where first I remove all files from this git repo and then added only necessary file.
> 
> Try to avoid this, please. We want to preserve the authoring of all our files that came from upstream.

@ricardozanini I had squash all my commit in this PR. Git history(authoring) of all files are back which I deleted and added back.

sutaakar commented:
I am not able to make a KogitoBuild running.
When I deploy example from https://github.com/vaibhavjainwiz/rhpam-kogito-operator/blob/KOGITO-4602/example/process-quarkus-example.yaml I get an error in operator:
```
021-03-18T13:05:57.433Z	ERROR	kogitoBuild-controller	Failed to update KogitoBuild	{"name": "example-quarkus", "namespace": "ksuta-test", "build_type": "source", "error": "KogitoBuild.rhpam.kiegroup.org \"example-quarkus\" is invalid: status.conditions: Required value"}
github.com/go-logr/zapr.(*zapLogger).Error
	/go/pkg/mod/github.com/go-logr/zapr@v0.1.1/zapr.go:128
github.com/kiegroup/kogito-operator/core/kogitobuild.(*statusHandler).HandleStatusChange
	/go/pkg/mod/github.com/kiegroup/kogito-operator@v0.12.1-0.20210316163042-b2f90df9984d/core/kogitobuild/status.go:80
github.com/kiegroup/rhpam-kogito-operator/controllers.(*KogitoBuildReconciler).Reconcile
	/workspace/controllers/kogitobuild_controller.go:109
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.6.3/pkg/internal/controller/controller.go:244
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.6.3/pkg/internal/controller/controller.go:218
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).worker
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.6.3/pkg/internal/controller/controller.go:197
k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1
	/go/pkg/mod/k8s.io/apimachinery@v0.18.8/pkg/util/wait/wait.go:155
k8s.io/apimachinery/pkg/util/wait.BackoffUntil
	/go/pkg/mod/k8s.io/apimachinery@v0.18.8/pkg/util/wait/wait.go:156
k8s.io/apimachinery/pkg/util/wait.JitterUntil
	/go/pkg/mod/k8s.io/apimachinery@v0.18.8/pkg/util/wait/wait.go:133
k8s.io/apimachinery/pkg/util/wait.Until
	/go/pkg/mod/k8s.io/apimachinery@v0.18.8/pkg/util/wait/wait.go:90
```

sutaakar commented:
KogitoRuntime seems to work.
I am ok with merging this PR and addressing the issues in next PRs.

ricardozanini commented:
> KogitoRuntime seems to work.
> I am ok with merging this PR and addressing the issues in next PRs.

Let's have @vaibhavjainwiz evaluate the fix. If it's a small thing, we can wait until tomorrow (my) morning.

spolti commented:
+1 for merging this.

vaibhavjainwiz commented:
@ricardozanini This issue has been caused because in `HandleStatusChange` method of KogitoBuild we are hard code creating condition object of `v1beta.KogitoBuildCondition` which is belong to community kogito-operator whereas it should be interface dependent. 
We had already fixed this scenario in KOGITO-4469, those PR is currently under review.
